Crash in DB update v18_to_v19

<!--## Intro

Use this template to report non-critical bugs. For critical bugs, anything that may involve loss or locking of funds, please submit the report to bounty@raiden.network.


## Issue checklist

To help us more easily debug your issue please check the boxes for which requirements for safe usage or Raiden were followed:

- [ ] The Ethereum node was always up to date with the chain tip and never stopped running.
- [x] No Ethereum transaction was sent for the account that Raiden manages by another app.
- [x] The Ethereum account used by Raiden always had sufficient ETH.
- [x] The Raiden DB that exists by default at ~/.raiden was never altered or deleted.
- [ ] The Raiden node was never restarted. If yes please provide more info.
- [x] The ethereum node was not switched while Raiden was running.
- [x] The Raiden REST API is protected from public access.

-->

## Problem Definition

I was running tests with the scenario player when nodes stopped starting (with reusing the token enabled).

During investigating I found the following log:
```
--------- Starting ---------
Traceback (most recent call last):
  File "/Users/paul/Projects/brainbot/raiden/venv/bin/raiden", line 11, in <module>
    load_entry_point('raiden', 'console_scripts', 'raiden')()
  File "/Users/paul/Projects/brainbot/raiden/raiden/__main__.py", line 11, in main
    run(auto_envvar_prefix='RAIDEN')  # pylint: disable=no-value-for-parameter
  File "/Users/paul/Projects/brainbot/raiden/venv/lib/python3.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/paul/Projects/brainbot/raiden/venv/lib/python3.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/paul/Projects/brainbot/raiden/venv/lib/python3.7/site-packages/click/core.py", line 1043, in invoke
    return Command.invoke(self, ctx)
  File "/Users/paul/Projects/brainbot/raiden/venv/lib/python3.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/paul/Projects/brainbot/raiden/venv/lib/python3.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/paul/Projects/brainbot/raiden/venv/lib/python3.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/paul/Projects/brainbot/raiden/raiden/ui/cli.py", line 494, in run
    app = runner.run()
  File "/Users/paul/Projects/brainbot/raiden/raiden/ui/runners.py", line 276, in run
    return self._start_services()
  File "/Users/paul/Projects/brainbot/raiden/raiden/ui/runners.py", line 95, in _start_services
    app_ = run_app(**self._options)
  File "/Users/paul/Projects/brainbot/raiden/raiden/ui/app.py", line 510, in run_app
    raiden_app.start()
  File "/Users/paul/Projects/brainbot/raiden/raiden/app.py", line 152, in start
    self.raiden.start()
  File "/Users/paul/Projects/brainbot/raiden/raiden/raiden_service.py", line 429, in start
    self.maybe_upgrade_db()
  File "/Users/paul/Projects/brainbot/raiden/raiden/raiden_service.py", line 1165, in maybe_upgrade_db
    manager.run()
  File "/Users/paul/Projects/brainbot/raiden/raiden/utils/upgrades.py", line 186, in run
    **self._kwargs,
  File "/Users/paul/Projects/brainbot/raiden/raiden/utils/upgrades.py", line 79, in _run_upgrade_func
    new_version = func(cursor, version, RAIDEN_DB_VERSION, **kwargs)
  File "/Users/paul/Projects/brainbot/raiden/raiden/storage/migrations/v18_to_v19.py", line 213, in upgrade_v18_to_v19
    _add_blockhash_to_state_changes(storage, cache)
  File "/Users/paul/Projects/brainbot/raiden/raiden/storage/migrations/v18_to_v19.py", line 70, in _add_blockhash_to_state_changes
    assert 'block_hash' not in data, 'v18 state changes cant contain blockhash'
AssertionError: v18 state changes cant contain blockhash
--------- Stopped ---------
```

And the logs:
```
{"raiden": "0.100.2rc5.dev610+g1a32bbd8.d20190326", "python_implementation": "CPython", "python_version": "3.7.2", "system": "macOS 10.14.4 64bit", "architecture": "x86_64", "distribution": "source", "event": "Starting Raiden", "logger": "raiden.ui.runners", "level": "info", "timestamp": "2019-03-26 13:55:11.719987"}
{"node": "1c6d403e", "available_nonce": 1, "client": "Geth/v1.8.22-stable/linux-amd64/go1.11.5", "event": "JSONRPCClient created", "logger": "raiden.network.rpc.client", "level": "debug", "timestamp": "2019-03-26 13:55:13.739154"}
{"available_servers": ["https://transport01.raiden.network", "https://transport02.raiden.network", "https://transport03.raiden.network"], "event": "Fetching available matrix servers", "logger": "raiden.ui.app", "level": "debug", "timestamp": "2019-03-26 13:55:14.733211"}
{"rtt_times": [["https://transport03.raiden.network", 0.020486666666666667], ["https://transport01.raiden.network", 0.02957966666666667], ["https://transport02.raiden.network", 0.10980899999999999]], "event": "Matrix homeserver RTT times", "logger": "raiden.network.transport.matrix.utils", "level": "debug", "timestamp": "2019-03-26 13:55:17.278880"}
{"sorted_servers": ["https://transport03.raiden.network", "https://transport01.raiden.network", "https://transport02.raiden.network"], "event": "Automatically selecting matrix homeserver based on RTT", "logger": "raiden.network.transport.matrix.utils", "level": "info", "timestamp": "2019-03-26 13:55:17.279289"}
{"event": "Upgrading database from 19 to v20", "logger": "raiden.utils.upgrades", "level": "debug", "timestamp": "2019-03-26 13:55:17.450880"}
{"event": "Failed to upgrade database: v18 state changes cant contain blockhash", "logger": "raiden.utils.upgrades", "level": "error", "timestamp": "2019-03-26 13:55:17.451883"}

``` 

### System Description

```
Darwin Pauls-MacBook-Pro-2.local 18.5.0 Darwin Kernel Version 18.5.0: Mon Mar 11 20:40:32 PDT 2019; root:xnu-4903.251.3~3/RELEASE_X86_64 x86_64
/usr/local/bin/solc
solc, the solidity compiler commandline interface
Version: 0.5.6+commit.b259423e.Darwin.appleclang
/usr/local/bin/geth
Geth
Version: 1.8.22-stable
Architecture: amd64
Protocol Versions: [63 62]
Network Id: 1
Go Version: go1.11.5
Operating System: darwin
GOPATH=
GOROOT=/usr/local/Cellar/go/1.11.5/libexec
/Users/paul/Projects/brainbot/raiden/venv/bin/raiden
{
  "raiden": "0.100.2rc5.dev610+g1a32bbd8.d20190326",
  "python_implementation": "CPython",
  "python_version": "3.7.2",
  "system": "macOS 10.14.4 64bit",
  "architecture": "x86_64",
  "distribution": "source"
}
```

--- a/raiden/tests/unit/storage/migrations/test_v16_to_v17.py
+++ b/raiden/tests/unit/storage/migrations/test_v16_to_v17.py
@@ -6,7 +6,7 @@ from unittest.mock import patch
 from raiden.storage.migrations.v16_to_v17 import upgrade_v16_to_v17
 from raiden.storage.sqlite import SQLiteStorage
 from raiden.tests.utils.migrations import create_fake_web3_for_block_hash
-from raiden.utils.upgrades import UpgradeManager
+from raiden.utils.upgrades import UpgradeManager, UpgradeRecord
 
 
 def setup_storage(db_path):
@@ -48,7 +48,7 @@ def test_upgrade_v16_to_v17(tmp_path):
     manager = UpgradeManager(db_filename=str(db_path), web3=web3)
     with patch(
             'raiden.utils.upgrades.UPGRADES_LIST',
-            new={16: upgrade_v16_to_v17},
+            new=[UpgradeRecord(from_version=16, function=upgrade_v16_to_v17)],
     ):
         manager.run()
 

--- a/raiden/tests/unit/storage/migrations/test_v16_to_v17.py
+++ b/raiden/tests/unit/storage/migrations/test_v16_to_v17.py
@@ -48,7 +48,7 @@ def test_upgrade_v16_to_v17(tmp_path):
     manager = UpgradeManager(db_filename=str(db_path), web3=web3)
     with patch(
             'raiden.utils.upgrades.UPGRADES_LIST',
-            new=[upgrade_v16_to_v17],
+            new={16: upgrade_v16_to_v17},
     ):
         manager.run()
 

--- a/raiden/tests/unit/storage/migrations/test_v17_to_v18.py
+++ b/raiden/tests/unit/storage/migrations/test_v17_to_v18.py
@@ -6,7 +6,7 @@ from unittest.mock import patch
 from raiden.storage.migrations.v17_to_v18 import upgrade_v17_to_v18
 from raiden.storage.sqlite import SQLiteStorage
 from raiden.tests.utils.migrations import create_fake_web3_for_block_hash
-from raiden.utils.upgrades import UpgradeManager
+from raiden.utils.upgrades import UpgradeManager, UpgradeRecord
 
 
 def setup_storage(db_path):
@@ -49,7 +49,7 @@ def test_upgrade_v17_to_v18(tmp_path):
 
     with patch(
             'raiden.utils.upgrades.UPGRADES_LIST',
-            new={17: upgrade_v17_to_v18},
+            new=[UpgradeRecord(from_version=17, function=upgrade_v17_to_v18)],
     ):
         manager.run()
 

--- a/raiden/tests/unit/storage/migrations/test_v17_to_v18.py
+++ b/raiden/tests/unit/storage/migrations/test_v17_to_v18.py
@@ -49,7 +49,7 @@ def test_upgrade_v17_to_v18(tmp_path):
 
     with patch(
             'raiden.utils.upgrades.UPGRADES_LIST',
-            new=[upgrade_v17_to_v18],
+            new={17: upgrade_v17_to_v18},
     ):
         manager.run()
 

--- a/raiden/tests/unit/storage/migrations/test_v18_to_v19.py
+++ b/raiden/tests/unit/storage/migrations/test_v18_to_v19.py
@@ -6,7 +6,7 @@ from unittest.mock import patch
 from raiden.storage.migrations.v18_to_v19 import upgrade_v18_to_v19
 from raiden.storage.sqlite import SQLiteStorage
 from raiden.tests.utils.migrations import create_fake_web3_for_block_hash
-from raiden.utils.upgrades import UpgradeManager
+from raiden.utils.upgrades import UpgradeManager, UpgradeRecord
 
 
 def setup_storage(db_path):
@@ -65,7 +65,7 @@ def test_upgrade_v18_to_v19(tmp_path):
     manager = UpgradeManager(db_filename=str(db_path), web3=web3)
     with patch(
             'raiden.utils.upgrades.UPGRADES_LIST',
-            new={18: upgrade_v18_to_v19},
+            new=[UpgradeRecord(from_version=18, function=upgrade_v18_to_v19)],
     ):
         manager.run()
 

--- a/raiden/tests/unit/storage/migrations/test_v18_to_v19.py
+++ b/raiden/tests/unit/storage/migrations/test_v18_to_v19.py
@@ -65,7 +65,7 @@ def test_upgrade_v18_to_v19(tmp_path):
     manager = UpgradeManager(db_filename=str(db_path), web3=web3)
     with patch(
             'raiden.utils.upgrades.UPGRADES_LIST',
-            new=[upgrade_v18_to_v19],
+            new={18: upgrade_v18_to_v19},
     ):
         manager.run()
 

--- a/raiden/tests/unit/storage/migrations/test_v19_to_v20.py
+++ b/raiden/tests/unit/storage/migrations/test_v19_to_v20.py
@@ -84,7 +84,7 @@ def test_upgrade_v19_to_v20(tmp_path):
     manager = UpgradeManager(db_filename=str(db_path), raiden=raiden_service_mock)
     with patch(
             'raiden.utils.upgrades.UPGRADES_LIST',
-            new=[upgrade_v19_to_v20],
+            new={19: upgrade_v19_to_v20},
     ):
         manager.run()
 

--- a/raiden/tests/unit/storage/migrations/test_v19_to_v20.py
+++ b/raiden/tests/unit/storage/migrations/test_v19_to_v20.py
@@ -9,7 +9,7 @@ from raiden.storage.sqlite import SerializedSQLiteStorage, SQLiteStorage
 from raiden.tests.utils.factories import make_32bytes, make_address
 from raiden.tests.utils.mocks import MockRaidenService
 from raiden.utils.serialization import serialize_bytes
-from raiden.utils.upgrades import UpgradeManager
+from raiden.utils.upgrades import UpgradeManager, UpgradeRecord
 
 
 def setup_storage(db_path):
@@ -84,7 +84,7 @@ def test_upgrade_v19_to_v20(tmp_path):
     manager = UpgradeManager(db_filename=str(db_path), raiden=raiden_service_mock)
     with patch(
             'raiden.utils.upgrades.UPGRADES_LIST',
-            new={19: upgrade_v19_to_v20},
+            new=[UpgradeRecord(from_version=19, function=upgrade_v19_to_v20)],
     ):
         manager.run()
 

--- a/raiden/tests/unit/test_upgrade.py
+++ b/raiden/tests/unit/test_upgrade.py
@@ -37,6 +37,7 @@ def setup_storage(db_path):
 
 
 def test_no_upgrade_executes_if_already_upgraded(tmp_path):
+    # Setup multiple old databases
     for version in [16, 17, 18, 19]:
         old_db_filename = tmp_path / Path(f'v{version}_log.db')
 
@@ -50,7 +51,44 @@ def test_no_upgrade_executes_if_already_upgraded(tmp_path):
 
     with patch('raiden.utils.upgrades._run_upgrade_func') as upgrade_mock:
         UpgradeManager(db_filename=db_path).run()
+        # Oldest database is of the same version as the current, no migrations should execute
         assert not upgrade_mock.called
+
+
+def test_upgrade_executes_necessary_migration_functions(tmp_path, monkeypatch):
+    old_db_filename = tmp_path / Path(f'v18_log.db')
+
+    storage = setup_storage(old_db_filename)
+    with patch('raiden.storage.sqlite.RAIDEN_DB_VERSION', new=18):
+        storage.update_version()
+        storage.conn.close()
+
+    db_path = tmp_path / Path('v20_log.db')
+
+    upgrade_functions = {}
+    for i in range(16, 20):
+        mock = Mock()
+        mock.return_value = i + 1
+        upgrade_functions[i] = mock
+
+    with monkeypatch.context() as m:
+        m.setattr(
+            raiden.utils.upgrades,
+            'UPGRADES_LIST',
+            upgrade_functions,
+        )
+        m.setattr(
+            raiden.utils.upgrades,
+            'RAIDEN_DB_VERSION',
+            19,
+        )
+
+        UpgradeManager(db_filename=db_path).run()
+
+    assert upgrade_functions[16].call_count == 0
+    assert upgrade_functions[17].call_count == 0
+    assert upgrade_functions[18].call_count == 1
+    assert upgrade_functions[19].call_count == 1
 
 
 def test_upgrade_manager_restores_backup(tmp_path, monkeypatch):

--- a/raiden/tests/unit/test_upgrade.py
+++ b/raiden/tests/unit/test_upgrade.py
@@ -36,6 +36,23 @@ def setup_storage(db_path):
     return storage
 
 
+def test_no_upgrade_executes_if_already_upgraded(tmp_path):
+    for version in [16, 17, 18, 19]:
+        old_db_filename = tmp_path / Path(f'v{version}_log.db')
+
+        storage = setup_storage(old_db_filename)
+
+        with patch('raiden.utils.upgrades.RAIDEN_DB_VERSION', new=version):
+            storage.update_version()
+            storage.conn.close()
+
+    db_path = tmp_path / Path('v19_log.db')
+
+    with patch('raiden.utils.upgrades._run_upgrade_func') as upgrade_mock:
+        UpgradeManager(db_filename=db_path).run()
+        assert not upgrade_mock.called
+
+
 def test_upgrade_manager_restores_backup(tmp_path, monkeypatch):
     db_path = tmp_path / Path('v17_log.db')
 

--- a/raiden/tests/unit/test_upgrade.py
+++ b/raiden/tests/unit/test_upgrade.py
@@ -6,7 +6,7 @@ from unittest.mock import ANY, Mock, patch
 import raiden.utils.upgrades
 from raiden.storage.sqlite import SQLiteStorage
 from raiden.tests.utils.migrations import create_fake_web3_for_block_hash
-from raiden.utils.upgrades import VERSION_RE, UpgradeManager, get_db_version
+from raiden.utils.upgrades import VERSION_RE, UpgradeManager, UpgradeRecord, get_db_version
 
 
 def test_version_regex():
@@ -65,11 +65,14 @@ def test_upgrade_executes_necessary_migration_functions(tmp_path, monkeypatch):
 
     db_path = tmp_path / Path('v20_log.db')
 
-    upgrade_functions = {}
+    upgrade_functions = []
     for i in range(16, 20):
         mock = Mock()
         mock.return_value = i + 1
-        upgrade_functions[i] = mock
+        upgrade_functions.append(UpgradeRecord(
+            from_version=i,
+            function=mock,
+        ))
 
     with monkeypatch.context() as m:
         m.setattr(
@@ -85,10 +88,10 @@ def test_upgrade_executes_necessary_migration_functions(tmp_path, monkeypatch):
 
         UpgradeManager(db_filename=db_path).run()
 
-    assert upgrade_functions[16].call_count == 0
-    assert upgrade_functions[17].call_count == 0
-    assert upgrade_functions[18].call_count == 1
-    assert upgrade_functions[19].call_count == 1
+    assert upgrade_functions[0].function.call_count == 0
+    assert upgrade_functions[1].function.call_count == 0
+    assert upgrade_functions[2].function.call_count == 1
+    assert upgrade_functions[3].function.call_count == 1
 
 
 def test_upgrade_manager_restores_backup(tmp_path, monkeypatch):
@@ -102,9 +105,14 @@ def test_upgrade_manager_restores_backup(tmp_path, monkeypatch):
         storage.update_version()
         storage.conn.close()
 
-    upgrade_functions = {16: Mock()}
+    upgrade_functions = [
+        UpgradeRecord(
+            from_version=16,
+            function=Mock(),
+        ),
+    ]
 
-    upgrade_functions[16].return_value = 17
+    upgrade_functions[0].function.return_value = 17
 
     web3, _ = create_fake_web3_for_block_hash(number_of_blocks=1)
     with monkeypatch.context() as m:
@@ -146,11 +154,14 @@ def test_sequential_version_numbers(tmp_path, monkeypatch):
 
     old_db_filename = tmp_path / Path('v16_log.db')
 
-    upgrade_functions = {16: Mock(), 17: Mock(), 18: Mock()}
-
-    upgrade_functions[16].return_value = 17
-    upgrade_functions[17].return_value = 18
-    upgrade_functions[18].return_value = 19
+    upgrade_functions = []
+    for i in range(16, 19):
+        mock = Mock()
+        mock.return_value = i + 1
+        upgrade_functions.append(UpgradeRecord(
+            from_version=i,
+            function=mock,
+        ))
 
     with patch('raiden.storage.sqlite.RAIDEN_DB_VERSION', new=16):
         storage = setup_storage(old_db_filename)
@@ -179,8 +190,8 @@ def test_sequential_version_numbers(tmp_path, monkeypatch):
 
         UpgradeManager(db_filename=db_path).run()
 
-        upgrade_functions[16].assert_called_once_with(ANY, 16, 19)
-        upgrade_functions[17].assert_called_once_with(ANY, 17, 19)
-        upgrade_functions[18].assert_called_once_with(ANY, 18, 19)
+        upgrade_functions[0].function.assert_called_once_with(ANY, 16, 19)
+        upgrade_functions[1].function.assert_called_once_with(ANY, 17, 19)
+        upgrade_functions[2].function.assert_called_once_with(ANY, 18, 19)
 
         assert get_db_version(db_path) == 19

--- a/raiden/tests/unit/test_upgrade.py
+++ b/raiden/tests/unit/test_upgrade.py
@@ -102,9 +102,9 @@ def test_upgrade_manager_restores_backup(tmp_path, monkeypatch):
         storage.update_version()
         storage.conn.close()
 
-    upgrade_functions = [Mock()]
+    upgrade_functions = {16: Mock()}
 
-    upgrade_functions[0].return_value = 17
+    upgrade_functions[16].return_value = 17
 
     web3, _ = create_fake_web3_for_block_hash(number_of_blocks=1)
     with monkeypatch.context() as m:
@@ -146,11 +146,11 @@ def test_sequential_version_numbers(tmp_path, monkeypatch):
 
     old_db_filename = tmp_path / Path('v16_log.db')
 
-    upgrade_functions = [Mock(), Mock(), Mock()]
+    upgrade_functions = {16: Mock(), 17: Mock(), 18: Mock()}
 
-    upgrade_functions[0].return_value = 17
-    upgrade_functions[1].return_value = 18
-    upgrade_functions[2].return_value = 19
+    upgrade_functions[16].return_value = 17
+    upgrade_functions[17].return_value = 18
+    upgrade_functions[18].return_value = 19
 
     with patch('raiden.storage.sqlite.RAIDEN_DB_VERSION', new=16):
         storage = setup_storage(old_db_filename)
@@ -179,8 +179,8 @@ def test_sequential_version_numbers(tmp_path, monkeypatch):
 
         UpgradeManager(db_filename=db_path).run()
 
-        upgrade_functions[0].assert_called_once_with(ANY, 16, 19)
-        upgrade_functions[1].assert_called_once_with(ANY, 17, 19)
-        upgrade_functions[2].assert_called_once_with(ANY, 18, 19)
+        upgrade_functions[16].assert_called_once_with(ANY, 16, 19)
+        upgrade_functions[17].assert_called_once_with(ANY, 17, 19)
+        upgrade_functions[18].assert_called_once_with(ANY, 18, 19)
 
         assert get_db_version(db_path) == 19

--- a/raiden/utils/upgrades.py
+++ b/raiden/utils/upgrades.py
@@ -17,12 +17,13 @@ from raiden.storage.sqlite import SQLiteStorage
 from raiden.storage.versions import VERSION_RE, older_db_file
 from raiden.utils.typing import Callable, Optional
 
-UPGRADES_LIST = [
-    upgrade_v16_to_v17,
-    upgrade_v17_to_v18,
-    upgrade_v18_to_v19,
-    upgrade_v19_to_v20,
-]
+# Source Target -> Upgrade Func
+UPGRADES_LIST = {
+    16: upgrade_v16_to_v17,
+    17: upgrade_v17_to_v18,
+    18: upgrade_v18_to_v19,
+    19: upgrade_v19_to_v20,
+}
 
 
 log = structlog.get_logger(__name__)

--- a/raiden/utils/upgrades.py
+++ b/raiden/utils/upgrades.py
@@ -15,15 +15,32 @@ from raiden.storage.migrations.v18_to_v19 import upgrade_v18_to_v19
 from raiden.storage.migrations.v19_to_v20 import upgrade_v19_to_v20
 from raiden.storage.sqlite import SQLiteStorage
 from raiden.storage.versions import VERSION_RE, older_db_file
-from raiden.utils.typing import Callable, Optional
+from raiden.utils.typing import Callable, NamedTuple, Optional
 
-# Source Target -> Upgrade Func
-UPGRADES_LIST = {
-    16: upgrade_v16_to_v17,
-    17: upgrade_v17_to_v18,
-    18: upgrade_v18_to_v19,
-    19: upgrade_v19_to_v20,
-}
+
+class UpgradeRecord(NamedTuple):
+    from_version: int
+    function: Callable
+
+
+UPGRADES_LIST = [
+    UpgradeRecord(
+        from_version=16,
+        function=upgrade_v16_to_v17,
+    ),
+    UpgradeRecord(
+        from_version=17,
+        function=upgrade_v17_to_v18,
+    ),
+    UpgradeRecord(
+        from_version=18,
+        function=upgrade_v18_to_v19,
+    ),
+    UpgradeRecord(
+        from_version=19,
+        function=upgrade_v19_to_v20,
+    ),
+]
 
 
 log = structlog.get_logger(__name__)
@@ -179,26 +196,14 @@ class UpgradeManager:
             try:
                 target_version = older_version
                 with storage.transaction():
-                    for source_version, upgrade_func in UPGRADES_LIST.items():
-                        # Source_version represents the version number
-                        # from which a migration starts to upgrade to a target version.
-                        # The target version is returned by the upgrade_func.
-                        # The below step basically skips a migration in case
-                        # the **latest older** version we have is larger than
-                        # the current migration's source version.
-                        # Example, assuming some node was running version 18 and
-                        # Upgraded to 20.
-                        # In this case, only migrations:
-                        # - 18_to_19
-                        # - 19_to_20
-                        # should run.
-                        if source_version < target_version:
+                    for upgrade_record in UPGRADES_LIST:
+                        if upgrade_record.from_version < target_version:
                             continue
 
                         target_version = _run_upgrade_func(
                             storage,
-                            upgrade_func,
-                            source_version,
+                            upgrade_record.function,
+                            upgrade_record.from_version,
                             **self._kwargs,
                         )
 

--- a/raiden/utils/upgrades.py
+++ b/raiden/utils/upgrades.py
@@ -75,10 +75,10 @@ def get_db_version(db_filename: Path) -> Optional[int]:
     return int(result[0])
 
 
-def _run_upgrade_func(cursor: sqlite3.Cursor, func: Callable, version: int, **kwargs) -> int:
+def _run_upgrade_func(storage: SQLiteStorage, func: Callable, version: int, **kwargs) -> int:
     """ Run the migration function, store the version and advance the version. """
-    new_version = func(cursor, version, RAIDEN_DB_VERSION, **kwargs)
-    update_version(cursor, new_version)
+    new_version = func(storage, version, RAIDEN_DB_VERSION, **kwargs)
+    update_version(storage, new_version)
     return new_version
 
 
@@ -177,13 +177,28 @@ class UpgradeManager:
             log.debug(f'Upgrading database from {older_version} to v{RAIDEN_DB_VERSION}')
 
             try:
+                target_version = older_version
                 with storage.transaction():
-                    version_iteration = older_version
-                    for upgrade_func in UPGRADES_LIST:
-                        version_iteration = _run_upgrade_func(
+                    for source_version, upgrade_func in UPGRADES_LIST.items():
+                        # Source_version represents the version number
+                        # from which a migration starts to upgrade to a target version.
+                        # The target version is returned by the upgrade_func.
+                        # The below step basically skips a migration in case
+                        # the **latest older** version we have is larger than
+                        # the current migration's source version.
+                        # Example, assuming some node was running version 18 and
+                        # Upgraded to 20.
+                        # In this case, only migrations:
+                        # - 18_to_19
+                        # - 19_to_20
+                        # should run.
+                        if source_version < target_version:
+                            continue
+
+                        target_version = _run_upgrade_func(
                             storage,
                             upgrade_func,
-                            version_iteration,
+                            source_version,
                             **self._kwargs,
                         )
 

